jbmc/src/java_bytecode/java_trace_validation.h

```diff
@@ -0,0 +1,41 @@
+/*******************************************************************\
+
+Module: Java trace validation
+
+Author: Jeannie Moulton
+
+\*******************************************************************/
+
+#ifndef CPROVER_JAVA_BYTECODE_JAVA_TRACE_VALIDATION_H
+#define CPROVER_JAVA_BYTECODE_JAVA_TRACE_VALIDATION_H
+
+#include <util/optional.h>
+
+class goto_tracet;
+class namespacet;
+class exprt;
+class symbol_exprt;
+
+/// Checks that the structure of each step of the trace matches certain
+/// criteria. If it does not, throw an error. Intended to be called by
+/// the caller of \ref build_goto_trace, for example
+/// \ref java_multi_path_symex_checkert::build_full_trace()
+void check_trace_assumptions(const goto_tracet &trace, const namespacet &ns);
+
+/// Returns true iff the expression is a symbol expression and has a non-empty
+/// identifier
+bool check_symbol_structure(const exprt &symbol_expr);
+/// Recursively extracts the first operand of an expression until it reaches a
+/// symbol and returns it, or returns an empty optional
+optionalt<symbol_exprt> get_inner_symbol_expr(exprt expr);
```

Comment:
It can seem strange that we would need a function that does that. 
I think we use it relying on the fact that dereference_exprt, typecast_exprt, index_exprt,  and byte_extract, all have the underlying object in the first operand. I think it should be checked in the function that we only go down expressions of that type, and document it here.

Reply:
OK, I've added a new commit that checks that the operand is one that we understand.